### DISCUSSION TITLE
Add a linter for `docs/`

### DISCUSSION THREAD

**Discussion by purva-thakre** [unitaryfund/mitiq#2440] [Ideas]
Previously discussed here: https://github.com/unitaryfund/mitiq/pull/2382#discussion_r1633795704

Option 1: There exists [a Sphinx extension for `.rst` files](https://github.com/sphinx-contrib/sphinx-lint). Will require additional configuration as most of our examples and user guides are written in `.md` format

Option 2: [`markdownlint`](https://github.com/DavidAnson/markdownlint) for `.md` files has a nice set of pre-defined [rules](https://github.com/DavidAnson/markdownlint/blob/main/doc/Rules.md). We can also define our own [rules](https://github.com/markdownlint/markdownlint/blob/41fc308f0d7f2647f0ae2c2d3826f48e42d964f4/docs/creating_rules.md) as needed. 

Option 3: https://github.com/jackdewinter/pymarkdown

**Comment by purva-thakre** [comment]
@natestemen What are your preferences?

I prefer Option 2 over the other two. The main caveat for this is that it is [not available as a PyPI package](https://github.com/DavidAnson/markdownlint?tab=readme-ov-file#related). We can use it in Github Actions or as a VS Code extension.

**Comment by purva-thakre** [comment]
Added a draft PR for this. 

We need to agree on which [rules](https://github.com/DavidAnson/markdownlint?tab=readme-ov-file#rules--aliases) we want to keep and which ones we want to ignore. The installation workflow for the cli is a little bit convoluted as I cannot use `pip` to install [`markdownlint-cli`](https://github.com/igorshubovych/markdownlint-cli). 

Locally, almost all the failures are due to violation of default line-length in the `.md` files in `docs/` after `markdownlint -f docs/` is used to fix the easily fixable errors. 

[markdownlint_all.txt](https://github.com/user-attachments/files/15894449/markdownlint_all.txt)
[markdownlint_minus_fixable_errors.txt](https://github.com/user-attachments/files/15894451/markdownlint_minus_fixable_errors.txt)



I did not want to commit the changes related to the easily fixable errors as we have not agreed on the ruleset yet. 

But once we agree on the rules, we can use the [`toml`](https://github.com/igorshubovych/markdownlint-cli?tab=readme-ov-file#configuration) file to ignore/follow a set of rules. 

https://github.com/DavidAnson/markdownlint/blob/b2305efafb034b1f328845aec9928b5363ffd646/schema/.markdownlint.yaml

https://github.com/unitaryfund/mitiq/blob/bf82e84410e1805437e15f1d26b0c5be8e8dcb8a/pyproject.toml#L1-L35

**Comment by jordandsullivan** [comment]
@purva-thakre can you add the new commit with the above edits? Let's try to figure out which issues this is actually fixing (i.e. formatting issues that would have been very obtrusive otherwise).

As well as deciding on rules.

**Comment by natestemen** [comment]
I left this comment over on [Purva's PR](https://github.com/unitaryfund/mitiq/pull/2411), but it more likely belongs here:

> I don't love the idea of adding node as a dependency to Mitiq. A simpler tool that can get us part of the way there is https://editorconfig.org/, and it's supported by almost all editors. Setting `trim_trailing_whitespace` to true would fix some of these issues. The drawback here is that this tool is not markdown specific.

**Comment by purva-thakre** [comment]
> The drawback here is that this tool is not markdown specific.

Since most of our docs pages are in `.md`, adding a tool that is not specific to markdown seems redundant. I tried adding a trailing whitespace to check if the tool flagged the errors. It did not. 

There are additional formatting issues with the `.md` files anyway. Some files have lines that run out of the page etc. `markdownlint` forces us to make changes to lines like this. 

How about we don't add `markdownlint` as a dependency or as a check to the `make format` lines? We can use it as an editor extension. The drawback to this is that it is not a compatible extension for all editors.

**Comment by purva-thakre** [comment]
Closing as the discussion is stale.
